File: Iter2/src/Transcript.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;

/**
 * The Transcript class that holds transcript information and contains related methods
 */
public class Transcript {
    
	/*
     * Multi-dimensional array to store the information of each course
     * Courses separated by semesters
     */
    private String[][][] transcript;
    
    /*
     * Cumulative GPA of all semesters
     */
    private double cumulativeGPA;
    
    /**
     * Array to store semester GPAs
     */
    private double[] GPAs;
    
    /*
     * Current semester number
     */
    private int currentSemester;
    
    /*
     * Reference owner of the transcript
     */
    private Student student;
    
    /*
     * Transcript creation date
     */
    private String creationDate;
    
    /*
     * Transcript JSON object
     */
    private JSONObject transcriptJSONObject;
    
    /*
     * Sum of taken credits in all semesters
     */
    private int cumulativeTakenCredits;
    
    /*
     * Sum of passed course credits in all semesters
     */
    private int cumulativeCompletedCredits;
    
    /*
     * By default, last semester will be assigned to current year
     */
    private final String currentYear;
   
    /*
     * Constructor of the Transcript
     */
    public Transcript(Student student, boolean includeNewCourses) {
        this.student = student;
        Date date = new Date();
        creationDate = new SimpleDateFormat("dd-MM-yyyy").format(date);
        currentYear = new SimpleDateFormat("yyyy").format(date);
        transcript = new String[student.getSemester()][student.getDepartment().getCourseTakeLimit()][5];
        GPAs = new double[student.getSemester()];
        currentSemester = student.getSemester();
        addCourses(student.getPreviousCourses());
        cumulativeGPA = 0;
        if (includeNewCourses) {
            addCourses(student.getNewlyTakenCourses());
        }
        transcriptJSONObject = createTranscript();
    }

    /*
     * Insert given courses into the transcript
     */
    public void addCourses(HashMap<DepartmentCourse, StudentCourse> courses) {
        for (DepartmentCourse departmentCourse : courses.keySet()) {
            StudentCourse course = courses.get(departmentCourse);
            int semesterIndex = course.getMainCourse().getSemester() - 1;
            if (course.getMainCourse() instanceof ElectiveCourse) {
                semesterIndex = student.getElectiveSemester((ElectiveCourse) course.getMainCourse()) - 1;
            }
            if (semesterIndex >= transcript.length || semesterIndex < 0) {
                continue;
            }
            // Insert course to first empty slot in transcript
            int i = 0;
            while (transcript[semesterIndex][i][0] != null && i < transcript[semesterIndex].length - 1) {
                i++;
            }
            transcript[semesterIndex][i][0] = course.getMainCourse().getCourseId();
            transcript[semesterIndex][i][1] = course.getMainCourse().getCourseName();
            transcript[semesterIndex][i][2] = String.valueOf(course.getMainCourse().getCredits());
            if (course.getRegisteredSemester() == currentSemester) {
                transcript[semesterIndex][i][3] = "";
                transcript[semesterIndex][i][4] = "";
            } else {
                transcript[semesterIndex][i][3] = course.getLetterGrade();
                transcript[semesterIndex][i][4] = String.valueOf(course.getPoint());
            }
        }
    }

    /*
     * Each transcript object is created using json objects and array.
     */
    private JSONObject createTranscript() {
        JSONObject transcriptObject = new JSONObject();
        JSONArray semesters = new JSONArray();
        for (int semesterNumber = 1; semesterNumber <= transcript.length; semesterNumber++) {
            semesters.add(createSemesterJSONObject(semesterNumber, semesterNumber == transcript.length));
        }
        transcriptObject.put("date", creationDate);
        transcriptObject.put("name", student.getName());
        transcriptObject.put("id", student.getStudentID());
        transcriptObject.put("level", student.getEducationLevel());
        transcriptObject.put("faculty", student.getFaculty());
        transcriptObject.put("department", student.getDepartment().getName());
        transcriptObject.put("semesters", semesters);
        return transcriptObject;
    }

    /*
     * Saves transcripts to the given filename
     */
    private void saveTranscript(JSONObject transcript, String outputFileName) {
        try (FileWriter file = new FileWriter(outputFileName)) {
            //We can write any JSONArray or JSONObject instance to the file
            file.write(transcript.toJSONString());
            file.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /*
     * A semester object is created that contains the number of semesters, the name of the semester,
     * the credits taken in the semester, the credits completed in the semester, the semester gpa,
     * the cumulative gpa, the cumulative credits taken, the cumulative completed credits and the course lists.
     */
    private JSONObject createSemesterJSONObject(int semesterNumber, boolean lastSemester) {
        JSONObject semesterObj = new JSONObject();
        JSONArray courseList = new JSONArray();
        int takenCredits = 0;
        int completedCredits = 0;
        double gpa = 0;
        String[][] semester = transcript[semesterNumber - 1];
        for (String[] courseDetails : semester) {
            if (courseDetails[0] != null) {
                int courseCredits = Integer.parseInt(courseDetails[2]);
                takenCredits += courseCredits;
                if (!(courseDetails[3].equals("FF") || courseDetails[3].equals("FD"))) {
                    completedCredits += courseCredits;
                }
                if (courseDetails[4].isEmpty()) {
                    courseDetails[4] = "0.0";
                }
                gpa += courseCredits * Double.parseDouble(courseDetails[4]);
                GPAs[semesterNumber - 1] = gpa;
                JSONObject course = createCourseJSONObject(courseDetails);
                courseList.add(course);
            }
        }
        gpa = roundGpa(gpa / takenCredits);
        cumulativeGPA += gpa;
        cumulativeTakenCredits += takenCredits;
        cumulativeCompletedCredits += completedCredits;
        semesterObj.put("semester", semesterNumber);
        semesterObj.put("semesterName", getSemesterName(semesterNumber) + getSemesterYear(semesterNumber));
        semesterObj.put("takenCredits", takenCredits);
        if (!lastSemester) {
            semesterObj.put("completedCredits", completedCredits);
            semesterObj.put("gpa", gpa);
            semesterObj.put("cumulativeGpa", roundGpa(cumulativeGPA / semesterNumber));
        }
        semesterObj.put("cumulativeTakenCredits", cumulativeTakenCredits);
        semesterObj.put("cumulativeCompletedCredits", cumulativeCompletedCredits);
        semesterObj.put("courses", courseList);
        return semesterObj;
    }

    /*
     * A result object is created that holds detailed course information.
     */
    private JSONObject createCourseJSONObject(String[] courseDetails) {
        JSONObject result = new JSONObject();
        for (int i = 0; i < courseDetails.length; i++) {
            result.put("courseId", courseDetails[0]);
            result.put("courseName", courseDetails[1]);
            result.put("credits", Integer.parseInt(courseDetails[2]));
            result.put("grade", courseDetails[3]);
            result.put("points", Double.parseDouble(courseDetails[4]));
        }
        return result;
    }

    /*
     * Find which year student attended to given semester
     * By default last semester in current year
     */
    private String getSemesterYear(int semesterNumber) {
        int currentYear = Integer.parseInt(this.currentYear);
        int semesterYear = currentYear - ((student.getSemester() - semesterNumber) / 2);
        return String.valueOf(semesterYear);
    }

    /*
     * gpa converter
     */
    private double roundGpa(double gpa) {
        return Math.round(100 * gpa) / 100.0;
    }

    /*
     * Gets semester number
     */
    private String getSemesterName(int semesterNumber) {
        return (semesterNumber % 2 == 0 ? "Spring" : "Fall");
    }

    /*
     * Gets transcript object
     */
    public JSONObject getTranscriptObject() {
        return transcriptJSONObject;
    }

    /*
     * Gets Transcript
     */
    public String[][][] getTranscript() {
        return transcript;
    }

    /**
     * Gets cumulative gpa
     */
    public double getCumulativeGPA() {
        return cumulativeGPA;
    }

    /*
     * Gets all gpas
     */
    public double[] getGPAs() {
        return GPAs;
    }

    /**
     * Gets Current Semester number
     */
    public int getCurrentSemester() {
        return currentSemester;
    }

    /*
     * Gets Student object
     */
    public Student getStudent() {
        return student;
    }

    /*
     * Gets Date of creation
     */
    public String getDateOfCreation() {
        return creationDate;
    }

    /*
     * Gets cumulative taken credits
     */
    public int getCumulativeTakenCredits() {
        return cumulativeTakenCredits;
    }

    /*
     * Gets cumulative completed credits
     */
    public int getCumulativeCompletedCredits() {
        return cumulativeCompletedCredits;
    }

    /*
     * Gets current Year
     */
    public String getCurrentYear() {
        return currentYear;
    }
}

File: Iter3/src/transcript.py
from datetime import datetime

import student


class Transcript:
    __transcript = []
    __cumulativeGpa = float
    __gpas = []
    __currentSemester = int
    __student = student
    __dateOfCreation = str
    __transcriptJSONObject = {}
    __cumulativeTakenCredits = int
    __cumulativeCompletedCredits = int
    __currentYear = str

    def __init__(self, student, includeNewCourses):
        self.__student = student
        self.__dateOfCreation = str(datetime.date(datetime.now))
        self.__currentYear = datetime.now.year
        self.__transcript = [student.get_semester()][student.get_department().get_course_take_limit()][5]
        self.__gpas = [student.get_semester]
        self.__cumulativeGpa = 0
        self.__currentSemester = student.get_semester()
        self.add_courses(student.get_previous_courses())
        if includeNewCourses == True:
            self.add_courses(student.get_newly_taken_courses())
        self.__transcriptJSONObject = self.create_transcript()

    def add_courses(self, courses):
        for departmentCourse in courses:
            course = courses(departmentCourse)
            semesterIndex = course.get_main_course().get_semester() - 1
            if isinstance(course.get_main_course, course.ElectiveCourse):
                semesterIndex = self.__student.get_elective_semester(course.ElectiveCourse(course.get_main_course())) - 1
            if semesterIndex >= len(self.__transcript) or semesterIndex < 0:
                continue
            i = 0
            while self.__transcript[semesterIndex][i][0] is None and i < len(self.__transcript[semesterIndex] - 1):
                i = i + 1
            self.__transcript[semesterIndex][i][0] = course.get_main_course().get_course_id()
            self.__transcript[semesterIndex][i][0] = course.get_main_course().get_course_name()
            self.__transcript[semesterIndex][i][2] = str(course.get_main_course().get_credits())
            if course.get_registered_semester() == self.__currentSemester:
                self.__transcript[semesterIndex][i][3] = ""
                self.__transcript[semesterIndex][i][4] = ""
            else:
                self.__transcript[semesterIndex][i][3] = course.get_letter_grade()
                self.__transcript[semesterIndex][i][4] = str(course.get_point())

    def create_transcript(self):
        transcriptObject = {}
        semesters = []
        for semesterNumber in range(1, len(self.__transcript)):
            semesters.append(self.create_semester_json_object(semesterNumber, semesterNumber is len(self.__transcript)))
        transcriptObject.append("date", self.__dateOfCreation)
        transcriptObject.append("name", self.__student.get_name())
        transcriptObject.append("id", self.__student.get_student_id())
        transcriptObject.append("level", self.__student.get_education_level())
        transcriptObject.append("faculty", self.__student.get_faculty())
        transcriptObject.append("department", self.__student.get_department().get_mame())
        transcriptObject.append("semesters", semesters)
        return transcriptObject

    def create_semester_json_object(self, semesterNumber, lastSemester):
        semesterObj = {}
        courseList = []
        takenCredits = 0
        completedCredits = 0
        gpa = 0
        semester = self.__transcript[semesterNumber - 1]

        for courseDetails in semester:
            if not courseDetails[0] is None:
                courseCredits = int(courseDetails[2])
                takenCredits += courseCredits
                if not courseDetails[3] is "FF" or courseDetails[3] is "FD":
                    completedCredits += courseCredits
                if courseDetails[4] is None:
                    courseDetails[4] = "0.0"
                gpa += courseCredits * float(courseDetails[4])
                self.__gpas[semesterNumber - 1] = gpa
                course = self.create_course_json_object(courseDetails[4])
                courseList.append(course)
        gpa = self.round_gpa(gpa / takenCredits)
        self.__cumulativeGpa += gpa
        self.__cumulativeTakenCredits += takenCredits
        self.__cumulativeCompletedCredits += completedCredits
        semesterObj.append("semester", semesterNumber);
        semesterObj.append("semesterName",
                           self.get_semester_name(semesterNumber) + self.get_semester_year(semesterNumber));
        semesterObj.append("takenCredits", takenCredits);
        if not lastSemester:
            semesterObj.append("completedCredits", completedCredits);
            semesterObj.append("gpa", gpa);
            semesterObj.append("cumulativeGpa", self.round_gpa(self.cumulativeGPA / semesterNumber));
        semesterObj.append("cumulativeTakenCredits", self.__cumulativeTakenCredits);
        semesterObj.append("cumulativeCompletedCredits", self.__cumulativeCompletedCredits);
        semesterObj.append("courses", courseList);
        return semesterObj

    def create_course_json_object(self, courseDetails):
        result = {}
        for i in range(0, len(courseDetails)):
            result.append("courseId", courseDetails[0]);
            result.append("courseName", courseDetails[1]);
            result.append("credits", int(courseDetails[2]));
            result.append("grade", courseDetails[3]);
            result.append("points", float(courseDetails[4]));
        return result

    def get_semester_year(self, semesterNumber):
        currentYear = int(self.__currentYear)
        semesterYear = currentYear - ((self.__student.get_semester() - semesterNumber) / 2)
        return str(semesterYear)

    def round_gpa(self, gpa):
        return round(100 * gpa) / 100.0

    def get_semester_name(self, semesterNumber):
        if semesterNumber % 2 == 0:
            return "Spring"
        return "Fall"

    def get_transcript_object(self):
        return self.__transcriptJSONObject

    def get_transcript(self):
        return self.__transcript

    def get_student(self):
        return self.__student
